twitter_api/main.py

```diff
@@ -1,23 +1,315 @@
+    if count == 0:
+        abort(404)
+    
+    # Find the user id of the user whose data match the submitted data
+    query = 'SELECT id FROM user WHERE username=? AND password=?;'
+    cursor = g.db.execute(query, (un, pw))
+    results = cursor.fetchone()
+    
+    # If there is an entry in the query result, store it as the user id
+    if results:
+        uid = results[0]
+    
+    # If the user fails authentication, return 401
+    else:
+        abort(401)
+    
+    # Generate an access token
+    ACCESS_TOKEN = _generate_access_token()
+    
+    # Store access token in *auth* table
+    query = 'INSERT INTO "auth" ("user_id", "access_token") VALUES (?, ?);'
+    g.db.execute(query, (int(uid), ACCESS_TOKEN))
+    g.db.commit()
+    response = Response(
+        json.dumps({'access_token': ACCESS_TOKEN}),
+        status=201,
+        mimetype=JSON_MIME_TYPE
+    )
+    
+    return response
+    
 
+@app.route('/logout', methods=['POST'])
+@auth_only
+def logout():
+    '''
+    Logs out the user by removing the user's access token from the auth table
+    '''
+    
+    # Remove the access token
+    query = 'DELETE FROM auth WHERE access_token=?;'
+    g.db.execute(query, (request.json['access_token'],))
+    g.db.commit()
+    
+    # Return a No Content response
+    return Response(status=204)
 
+@app.route('/profile', methods=['POST'])
+def update_profile():
+    
+    # Check if user is logged in. (user_id, access_token)
+    (user_id, db_token) = _check_login()
+    
+    # Verify that all required fields are supplied, and update the db
+    if all(_ in request.json for _ in {'first_name', 'last_name', 'birth_date'}):
+        
+        # Update new values in DB for those provided.
+        query = '''UPDATE user 
+                SET first_name=?, last_name=?, birth_date=? 
+                WHERE id=?;'''
+        fn = request.json['first_name']
+        ln = request.json['last_name']
+        dob = request.json['birth_date']
+        g.db.execute(query, (fn, ln, dob, user_id))
+        g.db.commit()
+        
+        # 201 - Success
+        return Response(status=201) 
+    
+    # 400 - Missing requirement, not json
+    return Response(status=400) 
+
+@app.route('/profile/<username>', methods=['GET'])
+def get_profile(username):
+    # Obtain existing data from user DB
+    (user_id, first_name, last_name, birth_date) = _get_existing_user_data(username)
+    user_data = {
+        'user_id': user_id,
+        'username': username,
+        'first_name': first_name,
+        'last_name': last_name,
+        'birth_date': birth_date
+    }
+    
+    # Obtain list of all user tweets by tweet-id
+    query = '''SELECT id FROM tweet WHERE user_id=?;'''
+    cursor = g.db.execute(query, (user_id,))
+    results = list(cursor.fetchall())
+    user_data['tweet'] = []
+    
+    for tweet_id in results:
+        (tweet_id, user_id, created, content) = _get_existing_tweet(tweet_id[0])
```

Comment:
You're making 1 query per tweet? I think you can make all this view work with just 1 SQL query.